pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -213,20 +246,212 @@ func (c *MCDefaultRouteController) processNextWorkItem() bool {
+				return initErr
+			}
+
+			if initErr = c.updateWireGuardPeers(); initErr != nil {
+				klog.ErrorS(initErr, "Failed to update WireGuard peers")
+				return initErr
+			}
+			c.wireGuardInitialized = true
+		}
+	} else {
+		ciImport, err := c.ciImportLister.ClusterInfoImports(c.namespace).Get(key)
+		if err != nil {
+			return err
+		}
+		return c.updateSingleWireGuardPeer(ciImport)
+	}
+
+	return nil
+}
+
+// updateWireGuardPeers lists all existing ClusterInfoImports and tries to add them as WireGuard peer.
+func (c *MCDefaultRouteController) updateWireGuardPeers() error {
+	klog.InfoS("Adding existing ClusterInfoImport as WireGuard peers")
+	ciImports, err := c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, ciImport := range ciImports {
+		if err := c.updateSingleWireGuardPeer(ciImport); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// updateSingleWireGuardPeer tries to update a WireGuard peer with ClusterInfoImport. If updating successfully,
+// it will also create host route the peer.
+func (c *MCDefaultRouteController) updateSingleWireGuardPeer(ciImport *mcv1alpha1.ClusterInfoImport) error {
+	klog.InfoS("Updating WireGuard peer with ClusterInfoImport", "ClusterInfoImport", ciImport)
+	if ciImport.Spec.WireGuard == nil || ciImport.Spec.WireGuard.PublicKey == "" {
+		klog.InfoS("ClusterInfoImport's WireGuard field has not been initialized, skip it", "ClusterInfoImport", ciImport)
+		return nil
+	}
+	serviceIP, serviceCIDR, err := net.ParseCIDR(ciImport.Spec.ServiceCIDR)
+	if err != nil {
+		return err
+	}
+	remoteWireGuardIP := serviceIP.Mask(serviceCIDR.Mask)
+	_, remoteWireGuardNet, err := net.ParseCIDR(remoteWireGuardIP.String() + "/32")
+	if err != nil {
+		return err
+	}
+
+	gatewayIP := net.ParseIP(ciImport.Spec.GatewayInfos[0].GatewayIP)
+	allowedIPs := []*net.IPNet{remoteWireGuardNet}
+	if err := c.wireGuardClient.UpdatePeer(ciImport.Name, ciImport.Spec.WireGuard.PublicKey, gatewayIP, 0, allowedIPs); err != nil {
+		klog.ErrorS(err, "Failed to update WireGuard peer", "ClusterInfoImport", ciImport)
+		return err
+	}
+	klog.InfoS("Updated WireGuard peer with ClusterInfoImport", "ClusterInfoImport", ciImport.Name)
+	if err := c.initWireGuardRoute(remoteWireGuardIP.String()); err != nil {
+		klog.ErrorS(err, "Failed to add WireGuard route on the host", "remoteGateway", ciImport.Name)
+		return err
+	}
+	return nil
+}
+
+// initializeWireGuard initializes the WireGuard interface and client.
+// It will also update Gateway's WireGuard field.
+func (c *MCDefaultRouteController) initializeWireGuard() error {
+	c.wireGuardConfig.MTU = c.nodeConfig.NodeTransportInterfaceMTU - c.networkConfig.MTUDeduction - config.WireGuardOverhead
+	wgClient, err := wireGuardNewFunc(c.nodeConfig, c.wireGuardConfig)
+	if err != nil {
+		return err
+	}
+	c.wireGuardClient = wgClient
+
+	clusterServiceCIDR, err := c.serviceCIDRProvider.GetServiceCIDR(false)
```

Comment:
Have you updated the code that sets the serviceCIDR in Gateway API? If no, the datapath won't work if the serviceCIDR in API and the discovered serviceCIDR don't match.

Can we get this cluster's own serviceCIDR from CIImport?

Reply:
No, we cannot get the serviceCIDR from CIImport, because controller won't import its own CIImport from the leader cluster. I will udpate the code  that sets the serviceCIDR in Gateway API, thanks.

Reply:
The patch still doesn't handle ServiceCIDR update case, the datapath will be broken as long as the discovered ServiceCIDR is updated.
How could user provide ServiceCIDR in MC configuration? I don't find an option in the configuration struct.
I think we shouldn't use unstructured data (annotations) in APIs specially designed for Antrea MultiCluster unless it's intended to be experimental and not be visible in API structure itself.

Reply:
How about this? If we WireGuard enabled, the Service CIDR must be provided in MulticlusterConfig?

Reply:
So, it is possible that (1st IP of) the discovered Service CIDR wont match the user provided Service CIDR? Because AntreaProxy may not discover the complete CIDR initially?

Reply:
@luolanzone ?

Reply:
@jianjuns  Yes, you are right,  that's why it's not suggested to use c.serviceCIDRProvider at the moment, I feel we can either warn user that a Service CIDR must be provided in MulticlusterConfig if WireGuard is enabled, or we create a local ClusterInfoImport to represent a member cluster itself, or a status struct in Gateway like Pod to include Service and Pod CIDR?

Reply:
I was thinking the `serviceCIDR` in MC controller configuration, not sure if that's proper to let antrea-agent to read the config?

Reply:
No. Probably let us add ServiceCIDR to Gateway CRD.

Reply:
@jianjuns So we can change the `Gateway` CDR like this ？
```bash
apiVersion: multicluster.crd.antrea.io/v1alpha1
gatewayIP: 172.18.0.4
internalIP: 172.18.0.4
# ADD a new field named "serviceCIDR"
serviceCIDR: 10.101.0.0/16
kind: Gateway
metadata:
  creationTimestamp: "2023-03-17T08:56:34Z"
  generation: 2
  name: east-worker
  namespace: kube-system
  resourceVersion: "1546"
  uid: 1df9c735-08d1-4e38-8ef2-b080041b0369
wireGuard:
  publicKey: uTqTl3dKZuuCpRncm9vHhrlemuuW1vbYVy/Awz2EJ2A=
```

When we need the local Service CIDR, we just need to get it from local Gateway.

Reply:
Yes, looks good to me.

Reply:
According to our discussion, I created another PR which stores ServiceCIDR in CRD Gateway. Could you please help to review? Thanks.
https://github.com/antrea-io/antrea/pull/4737/files